confidant/services/certificatemanager.py

```diff
@@ -0,0 +1,449 @@
+        of the issued certificate.
+        """
+        csr = csr.encode(encoding='UTF-8')
+        with stats.timer('issue_certificate'):
+            client = confidant.clients.get_boto_client('acm-pca')
+            response = client.issue_certificate(
+                CertificateAuthorityArn=self.settings['arn'],
+                Csr=csr,
+                SigningAlgorithm=self.settings['signing_algorithm'],
+                Validity={
+                    'Value': min(validity, self.settings['max_validity_days']),
+                    'Type': 'DAYS',
+                },
+                # Quick/easy idempotent token is just a hash of the csr itself.
+                # The token must be 36 chars or less.
+                IdempotencyToken=hashlib.sha256(csr).hexdigest()[:36],
+            )
+            return response['CertificateArn']
+
+    def _get_cached_certificate_with_key(self, cache_id):
+        """
+        For the cache id, get the cached response, or, if another thread is in
+        the process of issuing the same certificate, wait for the other thread
+        to populate the cache.
+        """
+        with stats.timer('get_cached_certificate_with_key'):
+            if not self.settings['certificate_use_cache']:
+                return {}
+            item = self.cache.get(cache_id)
+            # We're the first thread attempting to get this certificate
+            if not item:
+                return {}
+            # A certificate hasn't been issued yet, but since the cache id
+            # exists, another thread has requested the certificate.
+            i = 0
+            if not item.response:
+                # Wait for response
+                while True:
+                    # Only allow a maximum of 10s wait.
+                    # keep waiting while the lock is held.
+                    if item.lock:
+                        if i >= 100:
+                            break
+                        logging.debug(
+                            'Sleeping in _get_cached_certificate_with_key'
+                            ' for {}'.format(cache_id)
+                        )
+                        time.sleep(.100)
+                        i = i + 1
+                    else:
+                        break
+            # If the other thread failed to get the certificate, we need to
+            # ensure that this thread attempts to fetch a certificate.
+            return item.response
+
+    def issue_certificate_with_key(self, cn, validity, san=None):
+        """
+        Given the string common name, the validity length of the certificate (in
+        number of days), and a list of subject alternative names, return a dict
+        with the PEM encoded certificate, certificate chain, and private RSA
+        key.
+        """
+        with stats.timer('issue_certificate_with_key'):
+            cache_id = self.cache.get_cache_id(cn, validity, san)
```

Comment:
Do we need to add a check that caching is enabled?  Basically checking: `ACM_PRIVATE_CA_CERTIFICATE_USE_CACHE_`

Reply:
Hm. I guess the setting isn't really correct right now as we're still caching the data, but not delivering it. I'm going to make a no-op cache class and set that if it's unused.

Reply:
no biggy, just wanted to call it out.  can always follow up on an additional PR.

Reply:
Fixed